Update "format" and "content*" for new JSON Schema (#2200)
OAI/OpenAPI-Specification: versions/3.1.0.md

Note that `integer` as a type is also supported and is defined as a JSON number without a fraction or exponent part. 
Models are defined using the [Schema Object](#schemaObject), which is a superset of JSON Schema Specification Draft 2019-09.

<a name="dataTypeFormat"></a>Primitives have an optional modifier property: `format`, which is defined by JSON Schema.
OAS uses several known additional formats to define in fine detail the data type being used.
However, to support documentation needs, the `format` property is an open `string`-valued property, and can have any value.
Additional formats MAY be used even though undefined by either JSON Schema or this specification.
Types that are not accompanied by a `format` property follow the type definition in the JSON Schema. Tools that do not recognize a specific `format` MAY default back to the `type` alone, as if the `format` is not specified.

> Note that by default, JSON Schema validators MUST NOT attempt to validate the `format` keyword.  It is primarily treated as an annotation that informs applications of the intended nature of the data.  For historical reasons, it is possible to configure JSON Schema validators to validate "format", but this is not consistently implemented and SHOULD NOT be relied upon in any environment where an unknown validator or unknown validator configuration might be used.

Reviewer: In the company I work for we use UUIDs to identify resources so `type: string  format: uuid` is super common in our schemas. And we do leverage that information for a number of things. 
It seems that this note is motivated by the current implementations in the wild, which does not seem like a good reason to include it to a standard. If these validators start improving would this note go away? At what point? How many good validators out there would it make it happen?

Author: @jcarres-mdsol `uuid` and all other formats that are no longer explicitly listed are now part of JSON Schema proper.  See [Section 7.3.5 "Resource Identifiers"](https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.7.3.5) under the `format` keyword's specification for `uuid` and also how it relates to URNs of the form `urn:uuid:`.

Author: @jcarres-mdsol As for validators actually validating `format`, no, that's not going to happen.  I don't mean that in a "we (the JSON Schema project) decided it shouldn't happen", I mean it in "this has been a problem for years, and after talking with numerous maintainers of major implementations, it is 100% clear that `format` is a confusing mess that will not be validated consistently no matter how much the spec says it should be."  At some point you have to bow to reality and figure out a way forward, and saying that by default implementations MUST NOT attempt to validate using the `format` keyword was the only possible way to produce consistent default behavior.  This is [explained in detail](https://json-schema.org/draft/2019-09/release-notes.html#format-vocabulary) in the JSON Schema draft's release notes.

The problem with `format` is that it tries to do a huge number of things, *and* be open-ended to any extension anyone wants to throw in there.  There's not even agreement on what fully validating `format` would involve.  My hope is that with re-usable extension vocabularies being a thing, people will design vocabularies with keywords that validate specific things, instead of shoving them into this all-purpose-but-never-reliable keyword.

So an `isUuid` keyword would be straightforward to implement b/c UUIDs are well-specified.  For something like `email`, where there's no consensus on what "valid email" means, people could create keywords with specific criteria.  `isProbablyEmail` might just check for an `@` in the string (this is what several validators actually do right now).  `isEmailByRegex` might specify validation against a specific regex (instead of what happens now, which is that each validator comes up with its own incredibly complex regex that they probably got from a web site somewhere).  `isEmailBySomeLibraryName` would define what library to use to validate, etc.

These keywords would then be supported or not, which could be detected/enforced using the `$vocabulary` mechanism, and would reliably provide whatever level of validation they specify.  You could require support for `isUuid` but not require any sort of email validator, for example- that would be likely to be supported.  As would keywords for RFC 3339 date and time values, which are likewise well-specified.

Ultimately, I hope to deprecate `format` because it's always been a compatibility nightmare.  But that will only happen if enough replacement vocabularies are developed and adopted widely enough to be a viable migration path.

Reviewer: Thanks for the detailed response. Hopefully the plan will go as expected. I'm worried about the complexity of all this, I wonder if people will shy away and just keep shoveling everything into format anyways. But time will tell. 
Thanks for your efforts!

Author: @jcarres-mdsol we're definitely starting to see people understand the utility of the vocabulary concept on our slack.  But really, if enough people accept vocabularies as a solution here, we'll deprecate/remove `format`.  If they don't, we won't.

> Note that by default, JSON Schema validators MUST NOT attempt to validate the `format` keyword.  It is primarily treated as an annotation that informs applications of the intended nature of the data.  For historical reasons, it is possible to configure JSON Schema validators to validate "format", but this is not consistently implemented and SHOULD NOT be relied upon in any environment where an unknown validator or unknown validator configuration might be used.

Reviewer 2: Is there a reason for the double-quotes as opposed to backticks on the last mention of `"format"`?

Author: Nope, I'll fix that, thanks!

> Note that by default, JSON Schema validators MUST NOT attempt to validate the `format` keyword.  It is primarily treated as an annotation that informs applications of the intended nature of the data.  For historical reasons, it is possible to configure JSON Schema validators to validate `format`, but this is not consistently implemented and SHOULD NOT be relied upon in any environment where an unknown validator or unknown validator configuration might be used.

Reviewer 3: ```suggestion
Note that by default, JSON Schema validators will not attempt to validate the `format` keyword.  https://json-schema.org/draft/2019-09/release-notes.html#format-vocabulary
```

`integer` | `int64` | signed 64 bits (a.k.a long)
`number` | `float` | |
`number` | `double` | |
`string` | `password` | A hint to UIs to obscure input.
`string` | `byte` | base64 encoded characters **Deprecated**

> `string` | `binary` | any sequence of octets **Deprecated**

Reviewer 3: Should these simply be removed?  What would the impact be on other content-types?  If someone is using JSON Schema to infer schema on other non-JSON formats, what does format even mean?

Reviewer 3: Personally, I think we can remove these.

# content transferred with base64 encoding
schema:
  type: string
  contentEncoding: base64
```

```yaml
# content transferred in binary (octet-stream):
schema:
  type: string

>   contentMediaType: application/octet-stream

Reviewer 2: Should be include a `mediaTypeObject` in one of these examples? Is that the preferred OAS way to indicate media-type? What if there is a conflict between the `mediaTypeObject` key and a `contentMediaType` keyword?

Author: > What if there is a conflict between the `mediaTypeObject` key and a `contentMediaType` keyword?

So.... technically I would *not* use the schema in this example in the first place for a purely binary resource.  Because what this means is that there is a JSON string (double-quoted) with binary content, which doesn't actually make any sense.  You would have to encode the binary content into a legal JSON string (as indicated by `contentEncoding`).

For pure binary content, I would just us a Media Type Object without a schema.

Does that make sense?  I didn't want to start revamping larger things without first trying the minimal changes, as I assume this approach (`{"type": "string", "format": "binary"}`) has more or less worked for OAS so far.  But I wouldn't actually advise it, personally.

Author: @MikeRalphson @webron should I make the larger change here (to not use a Schema Object for binary content, and just use a Media Type Object)?

>       schema:

Reviewer 3: Should this schema be removed?

    'image/jpeg':

>       schema:

Reviewer 3: Schemas should be removed.
